### PR TITLE
Add the basic structure and some first rules for InnerSource repo lin…

### PR DESCRIPTION
**Pull request by ihrigb**
…ting.

### CODE REVIEW

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+public class GithubInnerSourceTopicDataProvider extends GitHubCachingDataProvider {
+
+  private static final Set<String> ACCEPTED_TOPICS;
+
+  static {
+    Set<String> acceptedTopics = new HashSet<>();
+    acceptedTopics.add("inner-source");
+    acceptedTopics.add("innersource");
+    ACCEPTED_TOPICS = Collections.unmodifiableSet(acceptedTopics);
+  }
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:27]:
```suggestion
  private static final Set<String> ACCEPTED_TOPICS = new HashSet<>(Arrays.asList("inner-source", "innersource"));
```

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package com.sap.oss.phosphor.fosstars.data.github;
+
+import static com.sap.oss.phosphor.fosstars.model.feature.oss.OssFeatures.HAS_INNERSOURCE_TOPIC;
+
+import com.sap.oss.phosphor.fosstars.model.Feature;
+import com.sap.oss.phosphor.fosstars.model.Value;
+import com.sap.oss.phosphor.fosstars.model.ValueSet;
+import com.sap.oss.phosphor.fosstars.model.other.Utils;
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:8]:
```suggestion
import static com.sap.oss.phosphor.fosstars.model.other.Utils.setOf;
```
Just to align with the other classes

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if (topicOptional.isPresent()) {
+      value = HAS_INNERSOURCE_TOPIC.value(true);
+    } else {
+      value = HAS_INNERSOURCE_TOPIC.value(false)
+        .explain("The repository does not have an InnerSource topic.");
+    }
+
+    return ValueHashSet.from(value);
+  }
+}
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:54]:
```suggestion
package com.sap.oss.phosphor.fosstars.data.github;

import static com.sap.oss.phosphor.fosstars.model.feature.oss.OssFeatures.HAS_INNERSOURCE_TOPIC;

import com.sap.oss.phosphor.fosstars.model.Feature;
import com.sap.oss.phosphor.fosstars.model.Value;
import com.sap.oss.phosphor.fosstars.model.subject.oss.GitHubProject;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

public class HasInnerSourceTopic extends CachedSingleFeatureGitHubDataProvider<Boolean> {

  private static final Set<String> ACCEPTED_TOPICS = new HashSet<>(Arrays.asList("inner-source", "innersource"));

  public HasInnerSourceTopic(GitHubDataFetcher fetcher) {
    super(fetcher);
  }

  @Override
  public Feature<Boolean> supportedFeature() {
    return HAS_INNERSOURCE_TOPIC;
  }

  @Override
  protected Value<Boolean> fetchValueFor(GitHubProject project) throws IOException {
    return HAS_INNERSOURCE_TOPIC.value(fetcher.githubTopicsFor(project).stream()
    .anyMatch(HasInnerSourceTopic::found))
    .explainIf(false, "The repository does not have an InnerSource topic.");
  }

  private static boolean found(String topic) {
    return ACCEPTED_TOPICS.contains(topic.toLowerCase(Locale.US));
  }
}
```
I have provided a gist of what we can use if only a single feature is present and this feature is extracted from GitHub

Also please add Java Docs and Logs where necessary maybe keep `HasSecurityTeam.java`

File: src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java [hunk 4]
```diff
@@ -555,4 +555,10 @@ private OssFeatures() {
    */
   public static final PositiveIntegerFeature NUMBER_OF_DEPENDENT_PROJECTS_ON_GITHUB
       = new PositiveIntegerFeature("Number of projects on GitHub that use an open source project");
+
+  /**
+   * Shows wether a Github repository has a Tag "inner-source".
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/feature/oss/OssFeatures.java:560]:
```suggestion
   * Shows whether a Github repository has the Tag "inner-source".
```

File: src/main/java/com/sap/oss/phosphor/fosstars/tool/format/CommonFormatter.java [hunk 5]
```diff
@@ -384,14 +393,4 @@ static String loadFrom(String resource, Class<?> clazz) {
...
-  /**
-   * Formats a double.
-   *
-   * @param n The double to be formatted.
-   * @return A formatted string.
-   */
-  static String formatted(double n) {
-    return DECIMAL_FORMAT.format(n);
-  }
 }
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/tool/format/CommonFormatter.java:396]: We need to add the feature to be displayed as part of the report in the final output, We need to add a line in the top of this file static block like `add(HAS_SECURITY_POLICY, "Does it have a security policy?");`

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
+    "java.compile.nullAnalysis.mode": "automatic",
+    "editor.tabSize": 2,
+    "java.configuration.updateBuildConfiguration": "interactive"
+}
```
Comment (sourabhsparkala) [inline .vscode/settings.json:5]: This content is specific to vscode right? Could we delete this, or add it to `.gitignore`

File: src/main/java/com/sap/oss/phosphor/fosstars/tool/format/CommonFormatter.java [hunk 7]
```diff
@@ -261,20 +261,17 @@ private static void add(Feature<?> feature, String caption) {
...
 
-  /**
-   * A formatter for doubles.
-   */
-  private static final DecimalFormat DECIMAL_FORMAT = new DecimalFormat("#.#");
-
   static {
-    DECIMAL_FORMAT.setMinimumFractionDigits(1);
-    DECIMAL_FORMAT.setMaximumFractionDigits(2);
   }
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/tool/format/CommonFormatter.java:265]: May I know why this was deleted and moved to a `constructor()`? :), we do not use this setting anywhere else.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java (hunk 1 above)
Comment (ihrigb) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:27]: No, `static final` would make (a) a class level variable and (b) a final reference to a Set. The Set itself would still be modifiable.

Comment (sourabhsparkala) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:27]: Yes you are right, I misread this. Thanks for clarifying.

**Comment by ihrigb** [comment]
Issue: #931

**Comment by ihrigb** [comment]
@sourabhsparkala as I updated everything already: any updates to this?

**Comment by sourabhsparkala** [comment]
@ihrigb I will look into this. In the mean time, there seems to be checkstyle issues, could you please address them?

Thanks
Sourabh

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  protected Value<Boolean> fetchValueFor(GitHubProject project) throws IOException {
+    Optional<String> topicOptional = fetcher.githubTopicsFor(project).stream().filter(topic -> {
+      return ACCEPTED_TOPICS.contains(topic.toLowerCase(Locale.US));
+    }).findFirst();
+
+    if (topicOptional.isPresent()) {
+      return HAS_INNERSOURCE_TOPIC.value(true);
+    }
+    return HAS_INNERSOURCE_TOPIC.value(false)
+        .explain("The repository does not have an InnerSource topic.");
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:45]:
```suggestion
boolean hasAcceptedTopic = fetcher.githubTopicsFor(project)
        .stream()
        .anyMatch(topic -> ACCEPTED_TOPICS.contains(topic.toLowerCase(Locale.US)));

    return HAS_INNERSOURCE_TOPIC.value(hasAcceptedTopic)
        .explainIf(false, "The repository does not have an InnerSource topic.");
  }
```

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import com.sap.oss.phosphor.fosstars.model.Feature;
+import com.sap.oss.phosphor.fosstars.model.Value;
+import com.sap.oss.phosphor.fosstars.model.subject.oss.GitHubProject;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/GithubInnerSourceTopicDataProvider.java:14]: Please add JAVA Docs to each method and variable
https://github.com/SAP/fosstars-rating-core/blob/master/src/main/java/com/sap/oss/phosphor/fosstars/data/github/IsApache.java

File: src/main/java/com/sap/oss/phosphor/fosstars/model/rating/oss/InnerSourceRulesOfPlayRating.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package com.sap.oss.phosphor.fosstars.model.rating.oss;
+
+import com.sap.oss.phosphor.fosstars.model.Confidence;
+import com.sap.oss.phosphor.fosstars.model.Label;
+import com.sap.oss.phosphor.fosstars.model.rating.AbstractRating;
+import com.sap.oss.phosphor.fosstars.model.score.oss.InnerSourceRulesOfPlayScore;
+import com.sap.oss.phosphor.fosstars.model.value.ScoreValue;
+import java.util.Objects;
+
+public class InnerSourceRulesOfPlayRating extends AbstractRating {
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/rating/oss/InnerSourceRulesOfPlayRating.java:10]: Please add JAVA Docs https://github.com/SAP/fosstars-rating-core/blob/master/src/main/java/com/sap/oss/phosphor/fosstars/model/rating/oss/OssRulesOfPlayRating.java#L11

File: src/main/java/com/sap/oss/phosphor/fosstars/model/score/oss/InnerSourceRulesOfPlayScore.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+import com.sap.oss.phosphor.fosstars.model.value.BooleanValue;
+import com.sap.oss.phosphor.fosstars.model.value.ScoreValue;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/score/oss/InnerSourceRulesOfPlayScore.java:19]: Please add the FEATURES  used for the score calculation as a JAVA Doc
As given here https://github.com/SAP/fosstars-rating-core/blob/master/src/main/java/com/sap/oss/phosphor/fosstars/model/score/oss/BanditScore.java#L15-L23

**Review by sourabhsparkala — CHANGES_REQUESTED** [review]
The changes look good. A few Java Docs are missing. Please note the below-required changes
- I believe you need to add entries of Features and Scores in the associated `*.java` files in the folder https://github.com/SAP/fosstars-rating-core/tree/master/src/main/java/com/sap/oss/phosphor/fosstars/tool/format
- Checkstyle checks failing in the GitHub Checks https://github.com/SAP/fosstars-rating-core/pull/924/checks

**Pull request closed without merging**